pkg/readiness/pruner/pruner_test.go

```diff
@@ -0,0 +1,185 @@
+	reg, err := wm.NewRegistrar(
+		cachemanager.RegistrarName,
+		events,
+	)
+	require.NoError(t, err, "creating registrar")
+
+	cfClient := &fakes.FakeCfClient{}
+	config := &cachemanager.Config{
+		CfClient:         cfClient,
+		SyncMetricsCache: syncutil.NewMetricsCache(),
+		Tracker:          tracker,
+		ProcessExcluder:  process.Get(),
+		Registrar:        reg,
+		Reader:           c,
+		GVKAggregator:    aggregator.NewGVKAggregator(),
+	}
+	require.NoError(t, err, "creating registrar")
+	cm, err := cachemanager.NewCacheManager(config)
+	require.NoError(t, err, "creating cachemanager")
+
+	if !startControllers {
+		// need to start the cachemanager if controllers are not started
+		// since the cachemanager is started in the controllers code.
+		require.NoError(t, mgr.Add(cm), "adding cachemanager as a runnable")
+	} else {
+		sw := watch.NewSwitch()
+		mutationSystem := mutation.NewSystem(mutation.SystemOpts{})
+
+		frameworksexternaldata.NewCache()
+		opts := controller.Dependencies{
+			CFClient:         testutils.SetupDataClient(t),
+			WatchManger:      wm,
+			ControllerSwitch: sw,
+			Tracker:          tracker,
+			ProcessExcluder:  process.Get(),
+			MutationSystem:   mutationSystem,
+			ExpansionSystem:  expansion.NewSystem(mutationSystem),
+			ProviderCache:    frameworksexternaldata.NewCache(),
+			CacheMgr:         cm,
+			SyncEventsCh:     events,
+		}
+		require.NoError(t, controller.AddToManager(mgr, &opts), "registering controllers")
+	}
+
+	testutils.StartManager(ctx, t, mgr)
+
+	return &ExpectationsPruner{
+		cacheMgr: cm,
+		tracker:  tracker,
+	}, c
+}
+
+// Test_ExpectationsMgr_DeletedSyncSets tests scenarios in which SyncSet and Config resources
+// get deleted after tracker expectations have been populated and we need to reconcile
+// the GVKs that are in the data client (via the cachemaanger) and the GVKs that are expected
+// by the Tracker.
+func Test_ExpectationsMgr_DeletedSyncSets(t *testing.T) {
+	tts := []struct {
+		name             string
+		fixturesPath     string
+		syncsetsToDelete []string
+		deleteConfig     string
+		startControllers bool
+	}{
+		{
+			name:             "delete all syncsets",
+			fixturesPath:     "testdata/syncsets-overlapping",
+			syncsetsToDelete: []string{"syncset-1", "syncset-2", "syncset-3"},
+		},
+		{
+			name:             "delete syncs and configs",
+			fixturesPath:     "testdata/syncsets-config-disjoint",
+			syncsetsToDelete: []string{"syncset-1"},
+			deleteConfig:     "config",
+		},
+		{
+			name:             "delete one syncset",
+			fixturesPath:     "testdata/syncsets-resources",
+			syncsetsToDelete: []string{"syncset-2"},
+			startControllers: true,
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancelFunc := context.WithCancel(context.Background())
+
+			require.NoError(t, testutils.ApplyFixtures(tt.fixturesPath, cfg), "applying base fixtures")
+
+			em, c := setupTest(ctx, t, tt.startControllers)
```

Comment:
I'm not sure I follow the thinking of this test.  The scenario we are trying to test is:

1. Tracker gets populated
2. Objects in the tracker's version of the world get deleted
3. Informers start and do not know about those objects (could be approximated by making sure controller never gets events for those objects)

Are we creating that scenario here?  Or are we approximating it with a different setup?

Reply:
Hm, not sure I understand, I think we have the same setup:

> Tracker gets populated

```
	require.Eventually(t, func() bool {
				return em.tracker.Populated()
			}, timeout, tick, "waiting on tracker to populate")
```

> Objects in the tracker's version of the world get deleted

```
for _, name := range tt.syncsetsToDelete {
...
}
```

> Informers start and do not know about those objects (could be approximated by making sure controller never gets events for those objects)

note that we don't start the controllers for most tests; when don't start the controllers no events are sent to the syncset_c or config_c for instance.

I've added a comment about this, let me know if it's still confusing.

Reply:
I added a test for a tracker with a separate lister to cover the superset scenario.